File: calendarapp/app/src/main/java/com/example/calendarapp/MainActivity.java
package com.example.calendarapp;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.util.Log;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.AdapterView;
import android.widget.GridView;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GestureDetectorCompat;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import java.util.Calendar;

public class MainActivity extends AppCompatActivity {
    private TextView tvYearMonth;
    private GridView gridCalendar;
    private Calendar calendar;
    private CalendarAdapter adapter;
    private GestureDetectorCompat gestureDetector;
    private BroadcastReceiver taskUpdateReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if ("com.example.calendarapp.TASK_UPDATED".equals(intent.getAction())) {
                // 只更新日曆視圖，不觸發其他操作
                adapter.notifyDataSetChanged();
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initializeViews();
        setupCalendar();
        setupAddButton();
        setupDateGridClick();
        setupGestureDetector();

        // 註冊廣播接收器
        IntentFilter filter = new IntentFilter("com.example.calendarapp.TASK_UPDATED");
        registerReceiver(taskUpdateReceiver, filter, Context.RECEIVER_NOT_EXPORTED);
    }

    private void initializeViews() {
        tvYearMonth = findViewById(R.id.tvYearMonth);
        gridCalendar = findViewById(R.id.gridCalendar);
    }

    private void setupCalendar() {
        calendar = Calendar.getInstance();
        updateCalendarView();
    }

    private void setupAddButton() {
        FloatingActionButton fabAdd = findViewById(R.id.fabAdd);
        fabAdd.setOnClickListener(v -> {
            Intent intent = new Intent(MainActivity.this, AddTaskActivity.class);
            startActivityForResult(intent, 1);
        });
    }

    private void setupDateGridClick() {
        gridCalendar.setOnItemClickListener((parent, view, position, id) -> {
            Calendar clickedDate = adapter.getDateAtPosition(position);
            if (clickedDate.get(Calendar.MONTH) == calendar.get(Calendar.MONTH)) {
                Intent intent = new Intent(MainActivity.this, TaskDetailActivity.class);
                intent.putExtra("year", clickedDate.get(Calendar.YEAR));
                intent.putExtra("month", clickedDate.get(Calendar.MONTH));
                intent.putExtra("day", clickedDate.get(Calendar.DAY_OF_MONTH));
                startActivity(intent);
            }
        });
    }

    private void setupGestureDetector() {
        gestureDetector = new GestureDetectorCompat(this, new SwipeGestureListener());
        gridCalendar.setOnTouchListener((v, event) -> gestureDetector.onTouchEvent(event));
    }

    private void updateCalendarView() {
        String yearMonth = String.format("%d年 %d月",
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH) + 1);
        tvYearMonth.setText(yearMonth);

        adapter = new CalendarAdapter(this, calendar);
        gridCalendar.setAdapter(adapter);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 1 && resultCode == RESULT_OK && data != null) {
            // 直接更新日曆視圖，不發送廣播
            if (adapter != null) {
                adapter.notifyDataSetChanged();
            }
        }
    }

    private class SwipeGestureListener extends GestureDetector.SimpleOnGestureListener {
        @Override
        public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
            if (Math.abs(velocityX) > Math.abs(velocityY)) {
                if (velocityX > 0) {
                    calendar.add(Calendar.MONTH, -1);
                } else {
                    calendar.add(Calendar.MONTH, 1);
                }
                updateCalendarView();
                return true;
            }
            return false;
        }
    }

    @Override
    protected void onDestroy() {
        unregisterReceiver(taskUpdateReceiver);
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/finalproject/MainActivity.kt
package com.example.finalproject

import android.content.Intent
import android.os.Bundle
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.floatingactionbutton.FloatingActionButton

class MainActivity : AppCompatActivity() {

    // 用來存儲待辦事項的清單
    private val todoList = mutableListOf<String>()
    private lateinit var todoAdapter: ArrayAdapter<String>

    // 用來顯示當前的月份和年份
    private lateinit var monthYearTextView: TextView

    companion object {
        private const val ADD_TODO_REQUEST_CODE = 1 // 用於辨識 AddTodoActivity 的返回結果
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // 設定月份年份的文字元件
        monthYearTextView = findViewById(R.id.monthYearTextView)
        monthYearTextView.text = "November 2024" // 預設月份年份

        // 初始化 ListView 和 Adapter
        val listView: ListView = findViewById(R.id.listView)
        todoAdapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, todoList)
        listView.adapter = todoAdapter

        // 左箭頭按鈕切換到前一個月
        val leftArrowButton: FloatingActionButton = findViewById(R.id.leftArrowButton)
        leftArrowButton.setOnClickListener {
            // TODO: 加入邏輯切換到前一個月
        }

        // 右箭頭按鈕切換到下一個月
        val rightArrowButton: FloatingActionButton = findViewById(R.id.rightArrowButton)
        rightArrowButton.setOnClickListener {
            // TODO: 加入邏輯切換到下一個月
        }

        // 浮動按鈕點擊事件，跳轉到 AddTodoActivity
        val fab: FloatingActionButton = findViewById(R.id.fab)
        fab.setOnClickListener {
            val intent = Intent(this, AddTodoActivity::class.java)
            startActivityForResult(intent, ADD_TODO_REQUEST_CODE)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        // 處理來自 AddTodoActivity 的返回結果
        if (requestCode == ADD_TODO_REQUEST_CODE && resultCode == RESULT_OK) {
            val todoText = data?.getStringExtra("TODO_TEXT") ?: return
            val todoDate = data?.getStringExtra("TODO_DATE") ?: return

            // 新增到清單並更新畫面
            addTodoToList(todoDate, todoText)
        }
    }

    // 新增待辦事項到清單
    private fun addTodoToList(date: String, todoText: String) {
        todoList.add("$date: $todoText")
        todoAdapter.notifyDataSetChanged() // 通知 Adapter 更新 UI
    }
}
